USB: Bluetooth over USB refactor and fixes

finikorg opened the pull request:
Move Bluetooth over USB to the new interface.

finikorg committed:
usb: Add BOS Descriptors

Add Binary Device Object Store descriptors

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Add sys_cpu_to_le16() conversion for USB field

Add missing conversion.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Add subsys/usb for device descriptor header

Support device descriptor header easy access

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: mass_storage: Use simpler header include

Use searched dirs for descriptor headers

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Add Bluetooth USB Device configuration options

Add USB_DEVICE_BLUETOOTH option

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Add Bluetooth device decriptors

Add USB Bluetooth device descriptors to usb_descriptor table.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Add Bluetooth device class core functionality

Implement Bluetooth over USB functionality through Bluetooth raw
access to the Bluetooth controller. Most devices with Bluetooth and
USB controllers supported by Zephyr can export themselves as USB
Bluetooth dongles.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Use new USB Device interface for Bluetooth over USB sample

Use new interface for the sample.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: hci_usb: Fix test name

Correct test name

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: hci_usb: Correct README

Correct README file

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

loicpoulain commented on subsys/usb/class/bluetooth.c:
```diff
@@ -0,0 +1,236 @@
+
+	/* Read number of bytes to read */
+	usb_read(ep, NULL, 0, &len);
+
```
We could use transfer API for reading a full HCI packet as well. Actually here we don't read a full HCI packet but a chunk of hci packet which is then transmitted to the Bluetooth driver via bt_send. In case of h4/UART (USB->Zephyr->UART->BTController), this works since the packet is retransmitted via UART and then re-aggregated on controller side, but I would expect that bt_send provides a complete HCI packet for larger compatibility.

finikorg replied:
I think I will try to use transfer for read also.

finikorg replied:
Now I think that usb_read() API is clear and much easy to understand then usb_transfer(READ), I leave it for now as is.

loicpoulain replied:
If it works no problem, we will potentially change it later.

finikorg commented on subsys/usb/class/bluetooth.c:
```diff
@@ -0,0 +1,236 @@
+		case BT_BUF_EVT:
+			usb_transfer_sync(CONFIG_BLUETOOTH_INT_EP_ADDR,
+					  buf->data, buf->len,
+					  USB_TRANS_WRITE | USB_TRANS_NO_ZLP);
```
For some reason without flag `USB_TRANS_NO_ZLP` we always send ZLP after each packet, is this desired behaviour?

loicpoulain replied:
> For some reason without flag USB_TRANS_NO_ZLP we always send ZLP after each packet, is this desired behaviour?
No this is not expected, we should only send ZLP when payload is a multiple of emps, to mark end of the transfer.

> I understand why ZLP is needed for ECM, since there is no packet length in ECM protocol and we mark end of the packet (and ONLY when packet is 64x bytes long), but why we need it for Bluetooth?
Actually, this is how a transfer is specified, the USB 'core' layer does not take care about payload (ECM, BT...), and needs a way to determine when a transfer is complete.

From USB spec:
"5.8.3 Bulk Transfer Packet Size Constraints
An endpoint must always transmit data payloads with a data field less than or equal to the endpoint’s
reported wMaxPacketSize value. When a bulk IRP involves more data than can fit in one maximum-sized data payload, all data payloads are required to be maximum size except for the last data payload, which will contain the remaining data. A bulk transfer is complete when the endpoint does one of the following:
• Has transferred exactly the amount of data expected
• Transfers a packet with a payload size less than wMaxPacketSize or transfers a zero-length packet"

finikorg replied:
Sorry, my bad, btw I have fixed issue with ZLP in PR #7110

finikorg committed:
usb: usb_dc_dw: Fix incorrect MPS return

Because of incorrectly set MPS we always got 0 returned for WRITE
transfers and ZLP were always generated for every packet.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

dbkinder commented on samples/bluetooth/hci_usb/README.rst:
```diff
@@ -13,8 +13,8 @@ board it runs on (e.g. :ref:`arduino_101` has this).
 
-* BlueZ running on the host, or
-* A board with BLE support
+* Bluetooth stack running on the host (e.g. Bluez)
```
change to ``BlueZ``

finikorg committed:
Merge branch 'usb-dw' into usb-bluetooth

dbkinder reviewed (approved):
Thanks!
